Delivery view test inconsistently failing in Travis

The bag delivery view test is inconsistently failing in Travis. It only fails when an initial build is started in Travis--such as with a new commit or PR. However, the test passes successfully when the build is restarted. Additionally, the test consistently passes locally.

```
FAIL: test_deliver_view (bagdiscovery.tests.BagTestCase)

Ensures BagDeliveryView executes BagDelivery routine.

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/local/lib/python3.6/unittest/mock.py", line 1183, in patched

    return func(*args, **keywargs)

  File "/code/bagdiscovery/tests.py", line 101, in test_deliver_view

    self.assertEqual(mock_post.call_count, 1)

AssertionError: 2 != 1

```

--- a/bagdiscovery/models.py
+++ b/bagdiscovery/models.py
@@ -12,10 +12,14 @@ class Bag(BasePackage):
     CREATED = 1
     DISCOVERED = 2
     DELIVERED = 3
+    DISCOVERING = 4
+    DELIVERING = 5
     PROCESS_STATUS_CHOICES = (
         (CREATED, "Created"),
         (DISCOVERED, "Discovered"),
-        (DELIVERED, "Delivered")
+        (DELIVERED, "Delivered"),
+        (DISCOVERING, "Discovering"),
+        (DELIVERING, "Delivering")
     )
     accession = models.ForeignKey(Accession, on_delete=models.CASCADE, null=True, blank=True)
     process_status = models.IntegerField(choices=PROCESS_STATUS_CHOICES, default=CREATED)

--- a/bagdiscovery/routines.py
+++ b/bagdiscovery/routines.py
@@ -19,8 +19,36 @@ class CleanupException(Exception):
     pass
 
 
-class BagDiscovery:
+class BaseRoutine(object):
+    """Base routine which contains main run method."""
+
+    def run(self):
+        if not Bag.objects.filter(process_status=self.in_process_status).exists():
+            bag = Bag.objects.filter(process_status=self.start_status).first()
+            if bag:
+                try:
+                    message = self.process_bag(bag)
+                except Exception as e:
+                    raise Exception(str(e), bag.bag_identifier)
+                bag.process_status = self.end_status
+                bag.save()
+            else:
+                message = self.idle_message
+        else:
+            message = "Service currently running"
+            bag = None
+        return (message, [bag.bag_identifier] if bag else None)
+
+    def process_bag(self, bag):
+        raise NotImplementedError("You must implement a process_bag method")
+
+
+class BagDiscovery(BaseRoutine):
     """Discovers and stores bags, and delivers data to another service."""
+    start_status = Bag.CREATED
+    in_process_status = Bag.DISCOVERING
+    end_status = Bag.DISCOVERED
+    idle_message = "No bags to discover."
 
     def __init__(self):
         self.src_dir = settings.SRC_DIR
@@ -38,20 +66,13 @@ class BagDiscovery:
             if not os.access(dir, os.W_OK):
                 raise BagDiscoveryException("Directory not writable", dir)
 
-    def run(self):
-        bag_ids = []
-        for bag in Bag.objects.filter(process_status=Bag.CREATED):
-            self.bag_name = "{}.tar.gz".format(bag.bag_identifier)
-            if os.path.exists(os.path.join(self.src_dir, self.bag_name)):
-                self.unpack_bag()
-                self.save_bag_data(bag)
-                self.move_bag(bag)
-                bag_ids.append(bag.bag_identifier)
-                bag.process_status = bag.DISCOVERED
-                bag.save()
-            else:
-                continue
-        return ("All bags discovered.", bag_ids)
+    def process_bag(self, bag):
+        self.bag_name = "{}.tar.gz".format(bag.bag_identifier)
+        if os.path.exists(os.path.join(self.src_dir, self.bag_name)):
+            self.unpack_bag()
+            self.save_bag_data(bag)
+            self.move_bag(bag)
+        return "All bags discovered."
 
     def unpack_bag(self):
         extracted = tar_extract_all(
@@ -100,26 +121,17 @@ class BagDiscovery:
                 bag.bag_identifier)
 
 
-class BagDelivery:
+class BagDelivery(BaseRoutine):
+    start_status = Bag.DISCOVERED
+    in_process_status = Bag.DELIVERING
+    end_status = Bag.DELIVERED
+    idle_message = "No bags to deliver."
 
-    def run(self):
-        bag = Bag.objects.filter(process_status=Bag.DISCOVERED).first()
-        if bag:
-            try:
-                self.deliver_data(bag, settings.DELIVERY_URL)
-                if bag.origin == "digitization":
-                    self.deliver_data(bag, settings.DERIVATIVE_DELIVERY_URL)
-                bag.process_status = Bag.DELIVERED
-                bag.save()
-                message = "All bag data delivered."
-            except Exception as e:
-                raise BagDiscoveryException(
-                    "Error sending metadata to {}: {}".format(
-                        settings.DELIVERY_URL, e))
-        else:
-            message = "No bags to deliver."
-            bag = None
-        return (message, [bag.bag_identifier] if bag else None)
+    def process_bag(self, bag):
+        self.deliver_data(bag, settings.DELIVERY_URL)
+        if bag.origin == "digitization":
+            self.deliver_data(bag, settings.DERIVATIVE_DELIVERY_URL)
+        return "All bag data delivered."
 
     def deliver_data(self, bag, url):
         try:

--- a/bagdiscovery/routines.py
+++ b/bagdiscovery/routines.py
@@ -72,7 +72,9 @@ class BagDiscovery(BaseRoutine):
             self.unpack_bag()
             self.save_bag_data(bag)
             self.move_bag(bag)
-        return "All bags discovered."
+            return "All bags discovered."
+        else:
+            raise Exception("Bag file does not exist.")
 
     def unpack_bag(self):
         extracted = tar_extract_all(

--- a/bagdiscovery/tests.py
+++ b/bagdiscovery/tests.py
@@ -64,13 +64,12 @@ class BagTestCase(TestCase):
         self.set_bag_status(Bag.CREATED)
         discovered = BagDiscovery().run()
         self.assertTrue(discovered)
-        for bag in Bag.objects.all():
-            self.assertEqual(bag.process_status, Bag.DISCOVERED)
-        self.assertEqual(len(listdir(settings.DERIVATIVE_CREATION_DIR)), 1)
+        self.assertTrue(Bag.objects.filter(process_status=Bag.DISCOVERED).exists())
 
     @patch('bagdiscovery.routines.requests.post')
     def test_deliver_bags(self, mock_post):
         """Ensures that BagDelivery routine runs without errors."""
+        mock_post.reset_mock()
         self.set_bag_status(Bag.DISCOVERED)
         message, bag_id = BagDelivery().run()
         self.assertEqual(message, "All bag data delivered.")
@@ -94,6 +93,7 @@ class BagTestCase(TestCase):
     @patch('bagdiscovery.routines.requests.post')
     def test_deliver_view(self, mock_post):
         """Ensures BagDeliveryView executes BagDelivery routine."""
+        mock_post.reset_mock()
         self.set_bag_status(Bag.DISCOVERED)
         request = self.factory.post(reverse('bagdelivery'))
         response = BagDeliveryView.as_view()(request)

--- a/bagdiscovery/tests.py
+++ b/bagdiscovery/tests.py
@@ -66,6 +66,13 @@ class BagTestCase(TestCase):
         self.assertTrue(discovered)
         self.assertTrue(Bag.objects.filter(process_status=Bag.DISCOVERED).exists())
 
+        shutil.rmtree(self.src_dir)
+        makedirs(self.src_dir)
+        self.set_bag_status(Bag.CREATED)
+        with self.assertRaises(Exception) as e:
+            BagDiscovery().run()
+        self.assertIn("Bag file does not exist.", str(e.exception))
+
     @patch('bagdiscovery.routines.requests.post')
     def test_deliver_bags(self, mock_post):
         """Ensures that BagDelivery routine runs without errors."""

